bundles/org.openhab.binding.flicbutton/src/main/java/org/openhab/binding/flicbutton/internal/FlicButtonHandlerFactory.java

```diff
@@ -0,0 +1,103 @@
+import org.openhab.core.thing.binding.ThingHandlerFactory;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link FlicButtonHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Patrick Fink - Initial contribution
+ */
+@Component(service = ThingHandlerFactory.class, configurationPid = "binding.flicbutton")
+public class FlicButtonHandlerFactory extends BaseThingHandlerFactory {
+
+    private final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Stream
+            .concat(FlicButtonBindingConstants.BRIDGE_THING_TYPES_UIDS.stream(),
+                    FlicButtonBindingConstants.SUPPORTED_THING_TYPES_UIDS.stream())
+            .collect(Collectors.toSet());
+    private final Map<ThingUID, @Nullable ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(FlicButtonBindingConstants.FLICBUTTON_THING_TYPE)) {
+            return new FlicButtonHandler(thing);
+        } else if (thingTypeUID.equals(FlicButtonBindingConstants.BRIDGE_THING_TYPE)) {
+            FlicButtonDiscoveryService discoveryService = new FlicSimpleclientDiscoveryServiceImpl(thing.getUID());
+            FlicDaemonBridgeHandler bridgeHandler = new FlicDaemonBridgeHandler((Bridge) thing, discoveryService);
+            registerDiscoveryService(discoveryService, thing.getUID());
+
+            return bridgeHandler;
+        }
+
+        return null;
+    }
+
+    @Override
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
+        if (thingHandler instanceof FlicDaemonBridgeHandler) {
+            unregisterDiscoveryService(thingHandler.getThing().getUID());
+        }
+        super.removeHandler(thingHandler);
+    }
+
+    private synchronized void registerDiscoveryService(FlicButtonDiscoveryService discoveryService,
+            ThingUID bridgeUID) {
+        this.discoveryServiceRegs.put(bridgeUID, getBundleContext().registerService(DiscoveryService.class.getName(),
+                discoveryService, new Hashtable<String, Object>()));
+    }
+
+    private synchronized void unregisterDiscoveryService(ThingUID bridgeUID) {
+        ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(bridgeUID);
+        if (serviceReg != null) {
+            FlicButtonDiscoveryService service = (FlicButtonDiscoveryService) getBundleContext()
+                    .getService(serviceReg.getReference());
+            if (service != null) {
+                service.deactivate();
+            }
+            serviceReg.unregister();
+            discoveryServiceRegs.remove(bridgeUID);
+        }
+    }
```

Comment:
You could use `ThingHandlerService` which would make this code unnecessary. The documentation for this is not yet merged, but available in the PR: https://github.com/openhab/openhab-docs/pull/1262/files#diff-c4a4d8725430bc2ea046182bfc73ac51349d989c7f6a8a6fa3001a226a09ad98R932

Reply:
@fwolter When I use this, the service (especially background discovery) seems to be activated completely independently from the bridge status which breaks the implementation and could (with the knowledge I currently have) only be fixed by implementing polling (either bridge polling for service or other way round). So is there an easy possibility to activate it when bridge is online and deactivate it again when bridge goes offline (as my current implementation does)?

Reply:
Why not finish the background task immediately if the bridge is offline?

Reply:
The problem is not going offline, but going online. I get a NPE because the background discovery starts automatically and tries to access bridge resources that are not available yet (flicClient).

Reply:
I think you have a general issue with null pointers as `FlicDaemonBridgeHandler` is missing the `@NonNullByDefault` annotation.

Reply:
Sure, null-safety could be improved, but that's not what this thread is about. The issue is that, when I use ThingHandlerService as you proposed in this thread, startBackgroundDiscovery() method is already called by the framework while Bridge is not initialized yet. If there is no easy solution for letting it start _after_ bridge is initialize, I would not use ThingHandlerService, because it does not provide what is needed to make it work.

Reply:
If you add the null annotation to `FlicDaemonBridgeHandler`, you will be forced to do a null check in the discovery for `flicClient`. That should prevent the NPE when using the `ThingHandlerService`. The background discovery would simply return, then.

Reply:
It's probably a timing issue. The discovery is started concurrently from setting the handler. This can result in the starting of the background to run before the setHandler. In general this is not a problem as most background discoveries run a scheduled task and this don't start right away. But here the background discovery is directly bound to the handler.
I think you can solve this by setting a boolean that background is started in the method `startBackgroundDiscovery` and check in `startBackgroundDiscovery` if handler is not null (set this boolean after handler check, and set to false in stop background discovery), In the sethandler, check if background is started and if yes call `startBackgroundDiscovery,`

Reply:
@fwolter Of course. But then the background discovery won't start because the needed steps are skipped. So yes, I would get rid of the NPE, but the functionality is not working anymore then, as you just ignore the problem by adding the null check and skipping the code you'd need to have executed. In fact, startBackgroundDiscovery() _needs_ a flicClient to work (and it's only called once and not repeatedly as far as I know).
@Hilbrand The handler is set, so that's not the issue here (`setThingHandler` method is called _before_ `startBackgroundDiscovery`). The issue is that the initialize() method is not completed when backgroundDiscovery starts, which means that `flicClient` is still null when startBackgroundDiscovery() is called. Doing it in the setHandler method does not help, as it's also called before bridge initialization - and therefore, initialization of flicClient.

Reply:
Use CompletableFuture. By this way discovery service can wait for bridge connection. Cause your ThingHandlerService has access to handler linking is not a problem.

Reply:
@splatch: But is that solution really better than the current one? Why adding complexity like polling, waiting, notifying, additional threads etc. even though the current solution is working fine without that complexity? The proposal by @fwolter was to use ThingHandlerService to simplify things. But now it looks like its just making it more complicated. If there is no solution that works without additional threads, I would prefer to stay with the current implementation which is working completely fine, even though it (admittedly) involves some boilerplate code.

Reply:
@pfink `CompletableFuture` is reliable way to avoid polling and introduce a hook without diving into tricks with service registrations. If your bridge is fully initialized before discovery hits then you have no impact. When discovery runs faster its get blocked until bridge sets value for future.
There is no need to implement polling in your code. It works using basic callbacks. If you dislike a future you can also use supplier which will block discovery service until bridge completes initialization. It is concurrent execution of code and can be solved with parallel functionalities of SDK.
I am not in favor of `ThingHandlerServices`, I believe they have many drawbacks (usage of reflection API, poor documentation). At the same time this is part of the framework which solves some problem.

Reply:
Decision by @pfink @fwolter in https://github.com/openhab/openhab-addons/pull/9234#issuecomment-887041267 and https://github.com/openhab/openhab-addons/pull/9234#issuecomment-890325082: 

> we can merge the current discovery implementation as-is.

So I resolve this thread.